File: Programmers/당구 연습.py
# 원 쿠션을 할 때 동, 서, 남, 북으로 총 4가지의 쿠션이 발생한다.
# 각각의 모든 경우에서의 거리를 계산하고 그 중에서 최솟값을 반환한다.
# 당구공의 이동 거리를 계산할 때는 '처음 당구공의 위치와 쿠션과의 거리 + 마지막 당구공의 위치와 쿠션과의 거리'가 아닌
# 처음 혹은 마지막 당구공의 위치를 쿠션으로부터 선대칭 이동하여 '처음 당구공의 위치를 선대칭 이동시킨 위치와 마지막 당구공의 위치'로 계산한다.

maxv = float('inf')

def convert(x, y, sx, sy, m, n):
    one = (sx+x)**2+(y-sy)**2
    if y == sy and sx > x:
        one = maxv
        
    four = (sy-y)**2+(2*m-x-sx)**2
    if y == sy and x > sx:
        four = maxv
        
    two = (sy+y)**2+(sx-x)**2
    if x == sx and y < sy:
        two = maxv
    
    three = (sx-x)**2+(2*n-y-sy)**2
    if x == sx and y > sy:
        three = maxv
      
    return min(one, two, three, four)
    
def solution(m, n, startX, startY, balls):
    answer = []
    
    for i in balls:
        x = i[0]
        y = i[1]
        answer.append(convert(x, y, startX, startY, m, n))
    
    return answer

File: Programmers/당구 연습.c
#include <stdio.h>
#include <limits.h>

int convert(int x, int y, int sx, int sy, int m, int n) {
    int maxv = INT_MAX;

    int one = (sx + x) * (sx + x) + (y - sy) * (y - sy);
    if (y == sy && sx > x) {
        one = maxv;
    }

    int four = (sy - y) * (sy - y) + (2 * m - x - sx) * (2 * m - x - sx);
    if (y == sy && x > sx) {
        four = maxv;
    }

    int two = (sy + y) * (sy + y) + (sx - x) * (sx - x);
    if (x == sx && y < sy) {
        two = maxv;
    }

    int three = (sx - x) * (sx - x) + (2 * n - y - sy) * (2 * n - y - sy);
    if (x == sx && y > sy) {
        three = maxv;
    }

    int min_val = one;
    if (two < min_val) min_val = two;
    if (three < min_val) min_val = three;
    if (four < min_val) min_val = four;

    return min_val;
}

void solution(int m, int n, int startX, int startY, int balls[][2], int ball_count, int answer[]) {
    for (int i = 0; i < ball_count; i++) {
        int x = balls[i][0];
        int y = balls[i][1];
        answer[i] = convert(x, y, startX, startY, m, n);
    }
}
